aten/src/ATen/native/Math.h

```diff
@@ -611,108 +617,121 @@ static scalar_t lanczos_sum_expg_scaled(scalar_t x) {
     39916800.,
     0.
   };
-  return ratevl(x, lanczos_sum_expg_scaled_num,
-      sizeof(lanczos_sum_expg_scaled_num) / sizeof(lanczos_sum_expg_scaled_num[0]) - 1,
-      lanczos_sum_expg_scaled_denom,
-      sizeof(lanczos_sum_expg_scaled_denom) / sizeof(lanczos_sum_expg_scaled_denom[0]) - 1);
+  return ratevl<scalar_t, is_cuda>(
+    x,
+    lanczos_sum_expg_scaled_num,
+    sizeof(lanczos_sum_expg_scaled_num) / sizeof(lanczos_sum_expg_scaled_num[0]) - 1,
+    lanczos_sum_expg_scaled_denom,
+    sizeof(lanczos_sum_expg_scaled_denom) / sizeof(lanczos_sum_expg_scaled_denom[0]) - 1
+  );
 }
 
-template <typename scalar_t>
-static scalar_t _igam_helper_fac(scalar_t a, scalar_t x) {
+template <typename scalar_t, bool is_cuda=false>
+static C10_HOST_DEVICE scalar_t
+_igam_helper_fac(scalar_t a, scalar_t x) {
   // compute x^a * exp(-a) / gamma(a)
   // corrected from (15) and (16) in [igam2] by replacing exp(x - a) with
   // exp(a - x).
-
-  scalar_t ax, fac, res, num, numfac;
-  static scalar_t MAXLOG = std::is_same<scalar_t,double>::value ?
-    7.09782712893383996843E2 : 88.72283905206835;
-  static scalar_t EXP1 = 2.718281828459045;
-  static scalar_t lanczos_g = 6.024680040776729583740234375;
-
-  if (std::fabs(a - x) > 0.4 * std::fabs(a)) {
-    ax = a * std::log(x) - x - std::lgamma(a);
+  using acc_t = at::acc_type<scalar_t, is_cuda>;
+  constexpr acc_t HALF = acc_t{0.5};
+  constexpr acc_t LARGE = acc_t{200.0};
+  const acc_t MAXLOG = std::is_same<scalar_t,double>::value ?
+    acc_t{7.09782712893383996843E2} : acc_t{88.72283905206835};
+  const acc_t EXP1 = acc_t{2.718281828459045};
+  const acc_t lanczos_g = acc_t{6.024680040776729583740234375};
+  acc_t ax, fac, res, num, numfac;
+
+  if (::fabs(a - x) > acc_t{0.4} * ::fabs(a)) {
```

Comment:
I know this stuff comes from the existing cuda code, but does CI actually break if we use the `std::` versions here? It's a bit hacky because technically the C signatures aren't overloaded, though in practice it probably works fine.

Reply:
CI is fine with `std::`, I had just assumed to really benefit on the CUDA side, we'd want to change from `std::` to globals to allow for the overloads. My understanding is not deep, though, when it comes to optimizing for GPUs, and I definitely would defer to the experts.

Reply:
I believe either the CUDA or ROCm builds were failing at one point with `std::` but in normal C++ the namespaced versions should be preferred. So, if it builds I would use the `std::` version.